File: AIP_Backend/Models/Customer.cs
#nullable enable

using System.ComponentModel.DataAnnotations;

namespace AIPBackend.Models
{
    public class Customer
    {
        [Key]
        public int CustomerId { get; set; }

        [Required]
        [MaxLength(100)]
        public string CompanyName { get; set; } = string.Empty;

        [Required]
        [MaxLength(50)]
        public string CompanyNumber { get; set; } = string.Empty;

        [MaxLength(50)]
        public string? VatNumber { get; set; }

        [Required]
        [MaxLength(50)]
        public string Status { get; set; } = "active";

        [MaxLength(50)]
        public string? CustomerType { get; set; }

        [MaxLength(100)]
        public string? Region { get; set; }

        // Address fields
        [MaxLength(100)]
        public string? Building { get; set; }

        [MaxLength(200)]
        public string? Street { get; set; }

        [MaxLength(100)]
        public string? Village { get; set; }

        [MaxLength(100)]
        public string? Town { get; set; }

        [MaxLength(100)]
        public string? County { get; set; }

        [MaxLength(10)]
        public string? Postcode { get; set; }

        // Contact fields
        [MaxLength(20)]
        public string? ContactTitle { get; set; }

        [MaxLength(50)]
        public string? ContactForename { get; set; }

        [MaxLength(50)]
        public string? ContactSurname { get; set; }

        [MaxLength(100)]
        public string? ContactPosition { get; set; }

        [MaxLength(100)]
        [EmailAddress]
        public string? ContactEmail { get; set; }

        [MaxLength(20)]
        public string? ContactPhone { get; set; }

        // Page assignments stored as JSON
        [MaxLength(4000)]
        public string? PageAssignments { get; set; }

        // Audit Fields
        [Required]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        [Required]
        [MaxLength(450)]
        public string CreatedBy { get; set; } = string.Empty;

        public DateTime? UpdatedAt { get; set; }

        [MaxLength(450)]
        public string? UpdatedBy { get; set; }

        // Navigation Properties
        public virtual ICollection<UserCustomerAssignment> UserCustomerAssignments { get; set; } = new List<UserCustomerAssignment>();
        public virtual ICollection<Region> Regions { get; set; } = new List<Region>();
        public virtual ICollection<Site> Sites { get; set; } = new List<Site>();
        public virtual ICollection<CustomerPageAccess> CustomerPageAccesses { get; set; } = new List<CustomerPageAccess>();
    }
}

File: AIP_UI/src/types/customer.ts
export interface Address {
  building: string
  street: string
  village?: string
  town: string
  county: string
  postcode: string
}

export interface Contact {
  title: string
  forename: string
  surname: string
  position: string
  email: string
  phone: string
}

export interface ViewConfig {
  id: string
  customerId: string // Changed from number to string to match mock data
  customerType: CustomerType
  enabledPages: string[]
  createdAt: string
  updatedAt: string
}

export type CustomerType = 'retail' | 'static' | 'gatehouse' | 'mobile-patrol' | 'keyholding-alarm-response' | 'event'

export interface CustomerPageAssignment {
  pageId?: string // Added for array structure
  enabled: boolean
  customized: boolean
  lastModified: string
  modifiedBy: string
}

export interface Customer {
  id: string // Changed from number to string to match mock data
  companyName: string
  companyNumber: string
  vatNumber: string
  status: 'active' | 'inactive'
  customerType: CustomerType
  address: Address
  contact: Contact
  viewConfig: ViewConfig
  pageAssignments?: CustomerPageAssignment[] // Changed from Record to array to match mock data
  assignedOfficers?: string[]
  createdAt: string
  updatedAt: string
}

export interface Region {
  regionID: number
  fkCustomerID: number
  regionName: string
  regionDescription?: string
  recordIsDeletedYN: boolean
  dateCreated: string
  createdBy: string
  dateModified?: string
  modifiedBy?: string
  // Navigation properties
  customer?: Customer
  sites?: Site[]
}

export interface Site {
  siteID: number
  fkCustomerID: number
  fkRegionID: number
  coreSiteYN: boolean
  locationName: string
  locationType?: string
  sinNumber?: string
  buildingName?: string
  numberandStreet?: string
  villageOrSuburb?: string
  town?: string
  county?: string
  postcode?: string
  telephoneNumber?: string
  contractStartDate?: string
  contractEndDate?: string
  details?: string
  siteSurveyComplete?: string
  assignmentInstructionsIssued?: string
  riskAssessmentIssued?: string
  recordIsDeletedYN: boolean
  dateCreated: string
  createdBy: string
  dateModified?: string
  modifiedBy?: string
  // Navigation properties
  customer?: Customer
  region?: Region
}

export interface CustomerPage {
  id: string
  title: string
  description: string
  enabled: boolean
  requiredForTypes: string[]
  path: string
  readOnly: boolean
  category: 'activity' | 'reporting' | 'security' | 'support' | 'settings' | 'incidents' | 'satisfaction' | 'safety' | 'reports'
  icon: string
}

export type CustomerPageId = 
  | 'daily-activity'
  | 'incident-graph'
  | 'incident-report'
  | 'customer-satisfaction'
  | 'be-safe-be-secure'
  | 'site-visit-reports'
  | 'keyholding-logs'
  | 'event-briefings'

export interface CustomerViewConfig {
  id: string
  customerId: number
  customerType: CustomerType
  enabledPages: string[]
  createdAt: string
  updatedAt: string
}

export interface CustomerReportingAccess {
  customerId: number
  customerName: string
  customerType: CustomerType
  assignedOfficers: string[]
  availablePages: CustomerPage[]
  lastActivity: string
  totalIncidents: number
  totalReports: number
}

export interface CustomerWithRelations extends Customer {
  regions: Region[]
  sites: Site[]
  availablePages?: CustomerPage[]
  statistics?: {
    incidents: number
    reports: number
    lastIncident?: string
    activeIssues: number
  }
}

export interface CustomerResponse {
  success: boolean
  data: Customer
  message?: string
}

export interface CustomersResponse {
  success: boolean
  data: Customer[]
  message?: string
}

export interface RegionResponse {
  success: boolean
  data: Region
  message?: string
}

export interface RegionsResponse {
  success: boolean
  data: Region[]
  message?: string
}

export interface SiteResponse {
  success: boolean
  data: Site
  message?: string
}

export interface SitesResponse {
  success: boolean
  data: Site[]
  message?: string
}

export interface CustomerReportingResponse {
  success: boolean
  data: CustomerReportingAccess[]
  message?: string
} 